packages/core/src/client/hmr/index.ts

```diff
@@ -216,3 +172,79 @@ function tryApplyUpdates() {
     );
   }
 }
+
+let connection: WebSocket | null = null;
+let isConnected = false;
+const MAX_RETRIES = 100;
+
+// Establishing a WebSocket connection with the server.
+function connect() {
+  isConnected = false;
+  connection = new WebSocket(socketUrl);
+
+  connection.onopen = function () {
+    isConnected = true;
+    if (typeof console !== 'undefined' && typeof console.info === 'function') {
+      // Notify users that the HMR has successfully connected.
+      console.info('[HMR] connected.');
+    }
+  };
+
+  // Attempt to reconnect after disconnection
+  connection.onclose = function () {
+    if (typeof console !== 'undefined' && typeof console.info === 'function') {
+      console.info('[HMR] disconnected. Attempting to reconnect.');
+    }
+    isConnected = false;
+
+    let retry_count = 0;
+    const retryInterval = setInterval(() => {
+      retry_count++;
+
+      if (isConnected || retry_count > MAX_RETRIES) {
+        clearInterval(retryInterval);
+        return;
+      }
+      reconnect();
```

Comment:
Is reconnect called too many times?

I tried to stop the local dev server, and the page creates many many requests:

![image](https://github.com/web-infra-dev/rsbuild/assets/7237365/27e74b47-b8d6-4188-b009-216017623c38)

Reply:
Oh, I'll take a look and see what's going on.

Reply:
When a reconnection attempt fails, it also triggers a 'close' event, leading to repeated setInterval calls. I'll address this issue in the fix~